Source/Core/Core/PowerPC/TCPGecko.h

```diff
@@ -0,0 +1,117 @@
+// Copyright 2016 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <atomic>
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <SFML/Network.hpp>
+
+#include "Common/CommonTypes.h"
+
+class TCPGecko
+{
+public:
+    TCPGecko();
+    ~TCPGecko();
+
+    bool Initialize();
+    void Terminate();
+
+private:
+    bool GetAvailableSock();
+
+    void ClientThread();
+    void CommandThread();
+
+    // Only ever one server thread
+    void GeckoConnectionWaiter();
+
+    bool IsClientValid() const;
+    static void SetMemCheck(u32 startAddress, u32 endAddress, u32 type);
+
+private:
+    // Client for this server object
+    std::unique_ptr<sf::TcpSocket> client;
+    std::thread clientThread;
+    std::thread commandThread;
+    std::mutex  transfer_lock;
+
+    std::deque<u8> send_fifo;
+    std::deque<u8> recv_fifo;
+
+    int               client_count;
+    std::atomic<bool> client_running;
+
+    std::atomic<bool> process_commands;
+
+    std::atomic<bool> server_running;
+    std::thread       connectionThread;
+    std::mutex        connection_lock;
+
+    std::queue<std::unique_ptr<sf::TcpSocket>> waiting_socks;
+
+    template<typename T>
+    bool read_data(T& data);
+    template<typename T>
+    bool read_data(std::vector<T>& data);
+    bool read_data(u8* data, u64 size);
+
+    template<typename T>
+    bool write_data(const T& data);
+    template<typename T>
+    bool write_data(const std::vector<T>& data);
+    bool write_data(const u8* data, u64 size);
+
+private:
+    enum class WiiStatus : u8
+    {
+        Running,
+        Paused,
+        Breakpoint,
+        Loader,
+        Unknown
+    };
+
+    static const u32 packetsize = 0xF800;
+    static const u32 uplpacketsize = 0xF80;
+
+    static const u8 cmd_poke08 = 0x01;
```

Comment:
These seem only be used in a switch. Wouldn't it be better to be an enum (class)?